### PR TITLE
trace: Add input and output filters options

### PR DESCRIPTION
**Pull request by dormanze**
## Description
Add input and output filters options.

## Motivation and Context
Many times I want to filter requests that use too much rate of flow.

## How to test this PR?
mc admin trace --help
trace-help
mc admin scanner trace --help
admin-scanner-help
mc admin trace --filter-request --filter-size 1MB myminio
image
mc admin trace --filter-response --filter-size 1MB myminio 
image

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fmc%2fpull%2fNNNNN)

### CODE REVIEW

**Comment by abperiasamy** [comment]
Since response-threshold and output-threshold mean the same, it makes sense to establish a new convention for all the thresholds to allow future extensibility and articulation. How about --filter-size --filter-duration --filter-request --filter-response?
--filter-size=1MB in combination with --filter-request is equivalent of --input-threshold, but articulates better.

**Comment by dormanze** [comment]
> Since response-threshold and output-threshold mean the same, it makes sense to establish a new convention for all the thresholds to allow future extensibility and articulation. How about --filter-size --filter-duration --filter-request --filter-response? --filter-size=1MB in combination with --filter-request is equivalent of --input-threshold, but articulates better.
This is a great suggestion, Thank you for your advice.I will change my pr.

**Comment by dormanze** [comment]
> Since response-threshold and output-threshold mean the same, it makes sense to establish a new convention for all the thresholds to allow future extensibility and articulation. How about --filter-size --filter-duration --filter-request --filter-response? --filter-size=1MB in combination with --filter-request is equivalent of --input-threshold, but articulates better.

This is an excellent suggestion.I will change my pr.

**Comment by dormanze** [comment]
> > Since response-threshold and output-threshold mean the same, it makes sense to establish a new convention for all the thresholds to allow future extensibility and articulation. How about --filter-size --filter-duration --filter-request --filter-response? --filter-size=1MB in combination with --filter-request is equivalent of --input-threshold, but articulates better.
> 
> This is an excellent suggestion.I will change my pr.

@abperiasamy Please review my pr again.

**Review by klauspost — COMMENTED** [review]
@abperiasamy I think a better general approach would be to have a simple filter language.

`--filter="response_duration GT 1s OR response_size GT 1MB"` - simple comparisons where we can plug in all the variables that makes sense without cluttering the flags completely. No math or anything funny, just `variable (comparator) constant [AND|OR ...]`.

File: cmd/admin-trace.go [hunk 1]
```diff
@@ -262,12 +312,28 @@ func matchingOpts(ctx *cli.Context) (opts matchOpts) {
 		ms.val = strings.TrimPrefix(s, "!")
 		opts.reqHeaders = append(opts.reqHeaders, ms)
 	}
+	var e error
+	var requestSize, responseSize uint64
+	if ctx.Bool("filter-request") && ctx.String("filter-size") != "" {
```
[the remark this answers is not in the document]
Comment (dormanze) [reply cmd/admin-trace.go:317]: Thank you for your quick reply.From line 164 to line 168, we have checked this error.
image

File: cmd/admin-trace.go [hunk 2]
```diff
@@ -81,6 +77,22 @@ var adminTraceFlags = []cli.Flag{
...
+	cli.BoolFlag{
+		Name:  "filter-response",
+		Usage: "trace calls only with response bytes greater than this threshold, use with filter-size",
+	},
+	cli.BoolFlag{
+		Name:  "filter-duration",
+		Usage: "trace calls only with response duration greater than threshold, use with filter-size",
+	},
+	cli.StringFlag{
+		Name:  "filter-size",
```
[the remark this answers is not in the document]
Comment (dormanze) [reply cmd/admin-trace.go:93]: Thank you for your reply.I suggest using `response-duration` , What do you think?

**Review by klauspost — APPROVED** [review]
LGTM

**Review by harshavardhana — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

**Comment by dormanze** [comment]
@harshavardhana @vadmeste @klauspost  Please approve my PR, thank you very much.

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**